src/main/java/ru/spbstu/service/VariantService.java

```diff
@@ -0,0 +1,30 @@
+package ru.spbstu.service;
+
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import ru.spbstu.model.Variant;
+import ru.spbstu.reader.dto.VcfRecord;
+import ru.spbstu.repository.VariantRepository;
+
+import java.util.stream.Stream;
+
+@Service
+@RequiredArgsConstructor
+public class VariantService {
+
+    private final VariantRepository variantRepository;
+
+    public void upsert(final Stream<VcfRecord> vcfRecords,
+                       final String dbName) {
+        vcfRecords.map(vcfRecord -> Variant.from(vcfRecord, dbName))
+                .forEach(this::upsert);
+    }
+
+    public void upsert(final Variant variant) {
+        final var variantToSave = variantRepository.findVariant(
```

Comment:
if there is not such variant then varianToSave = variant and this means that you'll duplicate annotations.
If it is really upsert action then you should create new variant i guess.